### DISCUSSION TITLE
Run clair as a binary

### DISCUSSION THREAD

**Discussion by chakradharv** [quay/clair#1227] [Q&A] [answered]
Hi. How do I run Clair as binary fine without the docker? I have a requirement to run Clair as binary file. I can setup my own postgres. Could you please help with this?

**Comment by hdonnay** [comment, marked as answer]
Building and running the binary should be fine. It's configured identically: set the flags or environment variables to indicate the mode and the config file.

**Reply by cvangeti** [reply]
@hdonnay I can build this binary from this https://github.com/quay/clair, right? Are there any reference on how to run the binary and scan the Images using binary?

**Reply by hdonnay** [reply]
[This](https://quay.github.io/clair/howto/getting_started.html#starting-clair-in-combo-mode) touches on running clair directly, but yes, just `go build` the clair and clairctl binaries.

**Reply by cvangeti** [reply]
I download  https://github.com/quay/clair there is a Makefile but not sure which task to run to build the binary. I see make tasks for containers but not for binaries.

**Reply by hdonnay** [reply]
Just run `go build`: `go build ./cmd/...`

**Reply by cvangeti** [reply]
I tried `go build ./cmd/...` in `https://github.com/quay/clair` directory but no services started up, nothing happened. Then I did `go run main.go -conf "../clair.config.yaml.sample" -mode "combo"` in `clair/cmd/clair` directory it is throwing error.
```
# command-line-arguments
./main.go:37:12: undefined: ConfValue
./main.go:39:12: undefined: ConfMode
./main.go:79:11: undefined: Shutdown
```

**Reply by hdonnay** [reply]
you need to run the binaries the build creates

**Reply by cvangeti** [reply]
After I ran this ` go build ./cmd/...` no binaries created. Where will the binaries be created?

**Reply by hdonnay** [reply]
in the directory where the command is run

**Reply by cvangeti** [reply]
Sorry, I don't see any binaries create in the directory. I ran `go build ./cmd/...` in the `https://github.com/quay/clair`

```
clair   (main) $  ls
CHANGELOG.md		NOTICE			config			httptransport		middleware
CODEOWNERS		README.md		config.yaml.sample	indexer			notifier
DCO			ROADMAP.md		contrib			initialize		openapi.yaml
Dockerfile		book.toml		docker-compose.yaml	internal		vendor
Documentation		clair-error		go.mod			introspection
LICENSE			cmd			go.sum			local-dev
Makefile		code-of-conduct.md	health			matcher
clair   (main) $  go build ./cmd/...
clair   (main) $  ls
CHANGELOG.md		NOTICE			config			httptransport		middleware
CODEOWNERS		README.md		config.yaml.sample	indexer			notifier
DCO			ROADMAP.md		contrib			initialize		openapi.yaml
Dockerfile		book.toml		docker-compose.yaml	internal		vendor
Documentation		clair-error		go.mod			introspection
LICENSE			cmd			go.sum			local-dev
Makefile		code-of-conduct.md	health			matcher
```

**Reply by hdonnay** [reply]
Sorry, I forgot that `build` won't copy out binaries if there are multiple packages. `go build ./cmd/clair` should do it.

Note that all this is covered in the go documentation and the online help for the `go` tool.

**Reply by cvangeti** [reply]
could you please point me to the go documentation?

**Reply by cvangeti** [reply]
Now I could able to run the binaries and Postgres is populated with data. I am trying to index an image using this API `/indexer/api/v1/index_report` where the body is
```
 const inputBody = '{
  "hash": "sha256:fc84b5febd328eccaa913807716887b3eb5ed08bc22cc6933a9ebf82766725e3",
  "layers": [
    {
      "hash": "sha256:fc84b5febd328eccaa913807716887b3eb5ed08bc22cc6933a9ebf82766725e3",
      "uri": "https://storage.example.com/blob/2f077db56abccc19f16f140f629ae98e904b4b7d563957a7fc319bd11b82ba36\n",
      "headers": {
        "property1": [
          "string"
        ],
        "property2": [
          "string"
        ]
      }
    }
  ]
}';
```
Here what is the URI I should mention?

**Reply by cvangeti** [reply]
@hdonnay could you please help me with this?

**Reply by cvangeti** [reply]
Also, I am getting this error for the centos:7 image 
```
./clairctl report centos:7
2021/04/15 11:50:05 indexer error: failed to scan all layer contents: rpm: failed to untar: exit status 1
```

**Reply by hdonnay** [reply]
It looks like you have a trailing newline on the `uri` field, which almost certainly shouldn't be there.

An indexer error indicates something went wrong in the indexing process. The clair logs should have more information; that error reports something wrong when attempting to run `tar`, my guess would be an invalid input or perhaps a `tar` that doesn't understand the arguments `clair` is using.

**Reply by jovandeginste** [reply]
> An indexer error indicates something went wrong in the indexing process. The clair logs should have more information; that error reports something wrong when attempting to run `tar`, my guess would be an invalid input or perhaps a `tar` that doesn't understand the arguments `clair` is using.

Correct; this is the full error (from the logs); seems like `--delay-directory-restore` is the issue...

```
rpm/Scanner.Scan,
kind:package,
layer:sha256:...,
scanner:rpm,
state:CheckManifest,
version:v0.0.2,
dir:/tmp/rpmscanner.870006625,
cmd:[
tar,-xC,/tmp/rpmscanner.870006625,--exclude,dev,--exclude,.wh*,--delay-directory-restore
],
stderr:tar: unrecognized option: delay-directory-restore
BusyBox v1.32.1 () multi-call binary.

Usage: tar c|x|t [-ZzJjahmvokO] [-f TARFILE] [-C DIR] [-T FILE] [-X FILE] [--exclude PATTERN]... [FILE]...

Create, extract, or list files from a tar file

	c	Create
	x	Extract
	t	List
	-f FILE	Name of TARFILE ('-' for stdin/out)
	-C DIR	Change to DIR before operation
	-v	Verbose
	-O	Extract to stdout
	-m	Don't restore mtime
	-o	Don't restore user:group
	-k	Don't replace existing files
	-Z	(De)compress using compress
	-z	(De)compress using gzip
	-J	(De)compress using xz
	-j	(De)compress using bzip2
	-a	(De)compress based on extension
	-h	Follow symlinks
	-T FILE	File with names to include
	-X FILE	File with glob patterns to exclude
	--exclude PATTERN	Glob pattern to exclude
 ```

**Reply by jovandeginste** [reply]
Apologies; above error was because I built my own container. The official container (v4.1.2) gives this error:

```json
{
  "level": "error",
  "component": "rpm/Scanner.Scan",
  "kind": "package",
  "layer": "sha256:...",
  "manifest": "sha256:...",
  "scanner": "rpm",
  "state": "CheckManifest",
  "version": "v0.0.2",
  "dir": "/tmp/rpmscanner.679648756",
  "cmd": [
    "tar",
    "--extract",
    "--exclude",
    "dev",
    "--exclude",
    ".wh*",
    "--delay-directory-restore",
    "--no-same-owner",
    "--no-same-permissions"
  ],
  "stderr": "",
  "err": "signal: killed",
  "time": "2021-08-11T08:40:11Z",
  "message": "error extracting layer"
}

{
  "level": "error",
  "component": "internal/indexer/controller/Controller.Index",
  "manifest": "sha256:...",
  "state": "CheckManifest",
  "error": "failed to scan all layer contents: zip error: zip: not a valid zip file",
  "time": "2021-08-11T08:40:11Z",
  "message": "error during scan"
}
```

**Comment by jovandeginste** [comment]
Just saw v4.2.0 was released, tested with that image, and having success now.

**Reply by crozzy** [reply]
Thanks for the update @jovandeginste, any chance you could share (non-sensitive) information about the image/manifest you are scanning.

**Reply by jovandeginste** [reply]
It was based on an old (centos6) image; legacy app...

**Reply by hdonnay** [reply]
I think that error is also quay/claircore#434
